### PR TITLE
ci: skip ts or go tests based on changed files

### LINKED ISSUE #7147: do not run go/ts tests for PRs that only target docs
We have a condition that should prevent running unnecessary tests for PRs that only target docs.

https://github.com/coder/coder/blob/d3356fa48e6c1a05c10e5782613d28ed3493919d/.github/workflows/ci.yaml#L101

But still, our ci is running all checks.
See https://github.com/coder/coder/actions/runs/4711074673/jobs/8355224760?pr=7048 for example.

cc: @mafredri, @deansheather

### PR DESCRIPTION
**Pull request by deansheather**
Closes #7147

### CODE REVIEW

**Review by BrunoQuaresma — APPROVED** [review]
Thank you!

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -28,6 +28,65 @@ concurrency:
...
+            go:
+              - "**.sql"
+              - "**.go"
+              - "**.golden"
+              - "go.mod"
+              - "go.sum"
+              # Other non-Go files that may affect Go code:
+              - "Makefile"
+              - "**.sh"
+              - "site/static/error.html"
```
[the remark this answers is not in the document]
Comment (mafredri) [reply .github/workflows/ci.yaml:63]: Would it make sense to do negative match instead?

E.g.

```yaml
- "**"
- "!site"
- ...
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -28,6 +28,65 @@ concurrency:
...
+      - uses: actions/checkout@v3
+      # For pull requests it's not necessary to checkout the code
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          filters: |
+            all:
+              - "**"
+            docs:
+              - "docs/**"
```
Comment (matifali) [inline .github/workflows/ci.yaml:49]: you should add `./README.md` heer too
```suggestion
              - "docs/**"
              - "./README.md"
```

**Review by matifali — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -28,6 +28,65 @@ concurrency:
   cancel-in-progress: ${{ github.event_name == 'pull_request' }}
 
 jobs:
+  changes:
+    runs-on: ubuntu-latest
+    outputs:
+      docs-only: ${{ steps.filter.outputs.docs_count == steps.filter.outputs.all_count }}
+      sh: ${{ steps.filter.outputs.sh }}
```
Comment (coadler) [inline .github/workflows/ci.yaml:35]: `sh` doesn't seem to be used anywhere

**Review by coadler — APPROVED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mafredri) [reply .github/workflows/ci.yaml:63]: For those I think it should work to do the following:

```yaml
- "**"
- "!site"
- "site/*.go"
```

[the remark this answers is not in the document]
Comment (deansheather) [reply .github/workflows/ci.yaml:63]: I'm kinda dubious whether that would work or not. There's no example in the readme of this action like that

[the remark this answers is not in the document]
Comment (deansheather) [reply .github/workflows/ci.yaml:63]: This doesn't seem to work in picomatch from my testing

**Review by matifali — APPROVED** [review]

**Comment by matifali** [comment]
unstale. @deansheather

**Review by ammario — CHANGES_REQUESTED** [review]
I tried this once before, but it caused issues because the required status checks would never pass on e.g. a docs PR.

We could get around that by checking the changed files in the `run` body, and `exit 0` early if no relevant files changed.

In an ideal world, test caching would make all of this a non-issue.

**Comment by matifali** [comment]
I discussed this with @deansheather and he said that he will change the settings to make the new required job as required. So it will work in that case.

**Review by ammario — APPROVED** [review]
Clever clever dean

**Comment by matifali** [comment]
I will wait for @deansheather to take a final look and then merge.

**Pull request merged**